### PR TITLE
[kmac/dv] initial version of KMAC smoke test

### PR DESCRIPTION
**Pull request by udinator**
This PR adds in the initial draft of the KMAC smoke test, with some additional enhancement commits needed to support it.

The first commit adds a function to the `str_utils_pkg` to convert a bytestring into a string representation.

The second commit updates some wording in the testplan on the condition under which we manually squeeze more output data.

The third commit updates the kmac DPI package to add function name and customization string inputs to the CSHAKE and KMAC functions.

Finally, the last commit adds the actual testbench code to implement the smoke test.

### CODE REVIEW

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (weicaiyang) [inline hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:347]: is this the same? if yes, we don't need line 341-345
```suggestion
    cfg_interrupts(.interrupts(intr));
```

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (weicaiyang) [inline hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:334]: nit: consider to use loop

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (weicaiyang) [inline hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:397]: use loop like this
https://github.com/lowRISC/opentitan/blob/c00e37f575a0eb5affb34c5e1901b1a266c90b70/hw/ip/keymgr/dv/env/seq_lib/keymgr_base_vseq.sv#L170-L176

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (weicaiyang) [inline hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:465]: seems like we should use `DV_CHECK_MEMBER_RANDOMIZE_FATAL`

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:347]: I don't think it's the same, because `cfg_interrupts` takes in an array of width TL_DW, but my instantiated array is of width 3.

I think you're right though, I should be able to remove this function since it doesn't do much else but change the array size.

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:397]: ohh that's a very good point, I was trying to come up with some macro-based approach for this, but looping like what you do is definitely the best approach here.
Thanks for the suggestion!

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:334]: you're right, I'll update it to use a similar loop to what you linked to below

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:465]: good point, will update this accordingly.

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (cindychip) [inline hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:218]: Just a suggestion, maybe make 2040 a parameter?

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (cindychip) [inline hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:469]: Just a comment - I think the tl_access has a non_blocking option, https://github.com/lowRISC/opentitan/blob/master/hw/dv/sv/cip_lib/cip_base_vseq.sv#L102
Right now it defaults to blocking, so i believe there is no need to wait no outstanding.
But would be necessary if you set the non-blocking to 1.

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:218]: sure, I can do that (would probably be clearer that way) ! thanks for the suggestion

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:469]: yup you're absolutely right!
I intend to randomize blocking/nonblocking accesses in a future PR, so the call to `wait_no_outstanding_access` would become "relevant" then.
Let me comment out this function call for now with a TODO.

**Review by weicaiyang — APPROVED** [review]
LGTM

**Comment by udinator** [comment]
latest force push fixes failures in `csr_hw_reset` test.

File: hw/dv/sv/str_utils/str_utils_pkg.sv
Comment (sriyerg) [inline hw/dv/sv/str_utils/str_utils_pkg.sv:128]: String is inherently a dynamic array of bytes. You probably don't need a function for this - you can simply cast it to string. The following should work:
```systemverilog
    byte foo[] = {...};
    $display("%0s", string'(bytes)); 
```

You could perhaps try this out and confirm.

File: hw/ip/kmac/dv/env/kmac_env_pkg.sv
Comment (sriyerg) [inline hw/ip/kmac/dv/env/kmac_env_pkg.sv:29]: nit: An alternate approach is to add `KmacNumIntrs` as the last literal of `kmac_intr_e` (see UART for example), which has a couple of advantages.

Comment (udinator) [reply hw/ip/kmac/dv/env/kmac_env_pkg.sv:29]: what are some benefits of doing this? i'm not sure I fully understand how this would work

File: hw/dv/sv/str_utils/str_utils_pkg.sv
Comment (sriyerg) [reply hw/dv/sv/str_utils/str_utils_pkg.sv:128]: The `str_to_bytes` above is not needed either.

File: hw/ip/kmac/dv/env/kmac_env_pkg.sv
Comment (sriyerg) [reply hw/ip/kmac/dv/env/kmac_env_pkg.sv:29]: - Enum literal is an immutable const, hence logically equivalent to a localparam
- Looping through an enum is easy (use this literal as an exit condition). Without it, you will need to add a loop counter
- No need to adjust the value of the parameter when interrupts are added / removed (forgetting to do that adds a small debug overhead)
- Adds the ability to set an 'illegal' value to an enum variable

File: hw/dv/sv/str_utils/str_utils_pkg.sv
Comment (udinator) [reply hw/dv/sv/str_utils/str_utils_pkg.sv:128]: sure, will give this a try.

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:224]: good idea, thanks!

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:164]: ack

File: hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/seq_lib/kmac_base_vseq.sv:236]: sure, will do so. it seems like the `dv_utils_pkg` already has some endian functionality contained, so would be good to aggregate them all.

File: hw/dv/sv/str_utils/str_utils_pkg.sv
Comment (udinator) [reply hw/dv/sv/str_utils/str_utils_pkg.sv:128]: this approach doesn't seem like it works, seems like we'll still need to explicitly convert bytes <-> strings.

File: hw/ip/kmac/dv/env/kmac_env_pkg.sv
Comment (udinator) [reply hw/ip/kmac/dv/env/kmac_env_pkg.sv:29]: make sense!
I updated the enum to add this value at the very end.

**Comment by udinator** [comment]
private CI failing is due to CSR test failures, #4515 resolves these issues so will need to wait until that gets merged.

**Review by sriyerg — APPROVED** [review]
We can revisit the str bytes conversion later. The rest LGTM.

**Comment by udinator** [comment]
> We can revisit the str bytes conversion later. The rest LGTM.

sgtm, thanks for the review!

**Pull request merged**
